### ISSUE TITLE
filterFeatures always passes getTaskNFeats() as the value for nselect

### ISSUE THREAD

**Issue by annette987** [mlr-org/mlr#2723]
Is this line correct? : https://github.com/mlr-org/mlr/blob/b57049d9ae4b73b8de7d012d648e7cca9a6478bc/R/filterFeatures.R#L131-L132 

Shouldn't the argument nselect, passed to filterFeatures, be used here instead of getTaskNFeats:
` fval = generate.fv.data(task = task, method = method, 
   nselect = nselect, ...)$data `

The value for nselect is calculated at lines 95-100: https://github.com/mlr-org/mlr/blob/b57049d9ae4b73b8de7d012d648e7cca9a6478bc/R/filterFeatures.R#L95-L100
So shouldn't that value be passed to the filters? I realise that most of the filters don't use nselect, but they have the option of doing so. At the moment, nselect will always be set to the total number of features, not the number of features to be selected, so it is not serving any purpose. I would like to be able to use it correctly.

nselect is used here to select the features to be returned from filterFeatures: https://github.com/mlr-org/mlr/blob/b57049d9ae4b73b8de7d012d648e7cca9a6478bc/R/filterFeatures.R#L188-L198
But the filters themselves are not able to make use of this value.

**Comment by jakob-r** [comment]
Do you have any example where `filterFeatures` is not working as expected? 
The approach used here is to calculate the filter values for all features and then select the `nselect` top features.
The only `Filter` that actually uses `nselect` is `mRMRe::mRMR.classic(..., feature_count = nselect)`. I don't know if it makes a difference there if you first calculate all the values and select afterwards or if you directly pass `feature_count`.

Summing up I would say this is no bug, just bad design.

**Comment by pat-s** [comment]
> So shouldn't that value be passed to the filters? I realise that most of the filters don't use nselect, but they have the option of doing so. At the moment, nselect will always be set to the total number of features, not the number of features to be selected, so it is not serving any purpose. I would like to be able to use it correctly.

Some filters support partial scoring. We implemented this in {mlr3filters}. Here is is not really used in the right context, you are right.
I also mentioned this a while ago in a different issue.

However, with `cache = TRUE` it does not make a difference here in runtime since we just calculate the score once and then reuse it from the cache.

AFAIR {praznik} filters are the only ones which support partial scoring.

**Comment by annette987** [comment]
I don't have an example (yet) of filterFeatures not working properly, although I was getting very poor results with mRMRe, so perhaps that could be why. But I am writing a custom filter that needs nselect set to the correct value.
Even if it is not considered a bug, but bad design, I would argue that it should be fixed, especially as it is such a simple fix. As it stands now, there is no point having the argument nselect as it has no effect. I'm happy to do the PR.

**Comment by pat-s** [comment]
It's debatable whether this is a bug or not, theoretically it does not do something "wrong" but just nothing. But that's not the point here :)

It should not have any effect on the "performance" of a filter, i.e. how the ranking is done.
Some filters create the ranking sequentially in a multivariate context, i.e. they are able to stop at a certain amount of features ranked.
Filters which are based on univariate rankings have to calculate all filter values first before the ranking can be done.

Happy to review a PR! But be aware that this involves quite some work possibly as you would have to check this ability for every filter. 
{praznik} filters also make use of `nselect`. 

In {mlr3filters} we do not use the {mRMRe} package anymore but only the implementation in the {praznik} package: https://github.com/mlr-org/mlr3filters/blob/fec63371c111ebbc8b5f399f3885b383b0dc53ca/R/FilterMRMR.R#L42-L46
Both packages produce slightly different results due to how the algorithm was implemented.

**Comment by pat-s** [comment]
Closing for now, feel free to open a PR when you have time @annette987.
